### PR TITLE
Implemented OpenLineageRunEventBuilder and defined OpenLineageEventHandlerFactory

### PR DESCRIPTION
**Pull request by collado-mike**
### Problem
Continuation of https://github.com/OpenLineage/OpenLineage/pull/447 . Defines the `OpenLineageEventHandlerFactory` interface, which must be implemented by extension authors, as well as the `OpenLineageRunEventBuilder` builder, which invokes the functions returned by the factory. This doesn't include the logic to actually find and instantiate the factories from the classpath. 

Much of the API interfaces an dbase classes are defined here as well.

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

### Checklist

- [X] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [ ] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)

### CODE REVIEW

File: integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java [hunk 1]
```diff
@@ -32,6 +33,7 @@
 @Value
 @Builder
 public class OpenLineageContext {
+  UUID runUuid = UUID.randomUUID();
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java:36]: _Minor_: Maybe define as `private final`?

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/InternalEventHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  private <T> List<T> generate(ServiceLoader<OpenLineageEventHandlerFactory> factories,
+      Function<OpenLineageEventHandlerFactory, List<T>> supplier) {
+    return StreamSupport.stream(
+            Spliterators.spliteratorUnknownSize(factories.iterator(), Spliterator.IMMUTABLE), false)
+        .flatMap(supplier.andThen(List::stream))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<PartialFunction<LogicalPlan, InputDataset>> createInputDatasetQueryPlanVisitors(
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/InternalEventHandlerFactory.java:76]: _Minor_: I would expect, generally, for factory methods to instantiate new objects each time invoked, but here we alway return the same object. Maybe not incorrect, but not what I'd expect.

**Review by wslulciuc — APPROVED** [review]
Only a minor comment, otherwise 👍

**Review by mobuchowski — APPROVED** [review]

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/UnknownEntryFacetListener.java [hunk 3]
```diff
@@ -4,21 +4,23 @@
...
 import java.util.stream.Collectors;
 import org.apache.spark.sql.catalyst.expressions.AttributeReference;
 import org.apache.spark.sql.catalyst.expressions.AttributeSet;
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
 import org.apache.spark.sql.types.DataType;
 import scala.collection.JavaConversions;
+import scala.runtime.AbstractPartialFunction;
 
 /**
  * Listener which is used in {@link PlanTraversal}. Holds information about visited nodes in {@link
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/UnknownEntryFacetListener.java:26]: Nevermind, this is handled in #451

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/InternalEventHandlerFactory.java (hunk 2 above)
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/InternalEventHandlerFactory.java]: I addressed this in https://github.com/OpenLineage/OpenLineage/pull/451

**Pull request merged**